Use mem3 in couch_multidb_changes to discover _replicator shards

nickva opened the pull request:
This is a forward-port of a corresponding commit in master:

"Use mem3 to discover all _replicator shards in replicator manager"

https://github.com/apache/couchdb-couch-replicator/commit/b281d2bb320ed6e6d8226765315a40637ba91a46

This wasn't a direct merge as replicator shard discovery and traversal is slightly
different.

`couch_multidb_changes` is more generic and takes a db suffix and callback
module. So `<<"_replicator">>` is not hard-coded in multidb changes module.

`couch_replicator_manager` handles local `_replicator` db by directly
creating it and launching a changes feed for it. In the scheduling replicator
creation is separate from monitoring. The logic is handled in `scan_all_dbs`
function where first thing it always checks if there is a local db present
matching the suffix, if so a `{resume_scan, DbName}` is sent to main process.
Due to supervisor order by the time that code runs a local replicator db
will be created already.

COUCHDB-3277

nickva committed:
Use mem3 in couch_multidb_changes to discover _replicator shards

This is a forward-port of a corresponding commit in master:

"Use mem3 to discover all _replicator shards in replicator manager"

https://github.com/apache/couchdb-couch-replicator/commit/b281d2bb320ed6e6d8226765315a40637ba91a46

This wasn't a direct merge as replicator shard discovery and traversal is slightly
different.

`couch_multidb_changes` is more generic and takes a db suffix and callback
module. So `<<"_replicator">>` is not hard-coded in multidb changes module.

`couch_replicator_manager` handles local `_replicator` db by directly
creating it and launching a changes feed for it. In the scheduling replicator
creation is separate from monitoring. The logic is handled in `scan_all_dbs`
function where first thing it always checks if there is a local db present
matching the suffix, if so a `{resume_scan, DbName}` is sent to main process.
Due to supervisor order by the time that code runs a local replicator db
will be created already.

COUCHDB-3277

sagelywizard commented on src/couch_multidb_changes.erl:
```diff
@@ -648,22 +665,77 @@ t_start_link_no_ddocs() ->
+
+
+t_fail_local() ->
+ ?_test(begin
```
Spacing

sagelywizard commented on src/couch_multidb_changes.erl:
```diff
@@ -648,22 +665,77 @@ t_start_link_no_ddocs() ->
 
+scan_dbs_test_() ->
+{
+      foreach,
```
Spacing

nickva replied:
Will fix it

sagelywizard commented on src/couch_multidb_changes.erl:
```diff
@@ -249,34 +250,44 @@ changes_reader_cb(_, _, Acc) ->
-        end, 1).
+    ok = scan_local_db(Server, DbSuffix),
+    {ok, Db} = mem3_util:ensure_exists(
+        config:get("mem3", "shards_db", "_dbs")),
```
CouchDB is `dbs` DB but dbnext is `_dbs` DB?

sagelywizard replied:
And the config param is `shards_db` vs `shard_db`?

nickva replied:
Yap. In ASF CouchDB default is `_dbs` and in dbcore it was `dbs`. And `shards_db` used to be `shard_db` as well.  At Cloudant we use `dbs` in the default.ini still.

Initially when I made the PR for replicator's master branch I mistakenly forgot to update this code so it had dbcore defaults. So there is a second commit there to correct it. In the scheduler's case, I just made it one commit.

sagelywizard commented on src/couch_multidb_changes.erl:
```diff
@@ -249,34 +250,44 @@ changes_reader_cb(_, _, Acc) ->
-    Range = min(2 * N * ?AVG_DELAY_MSEC, ?MAX_DELAY_MSEC),
-    random:uniform(Range).
+scan_local_db(Server, DbSuffix) when is_pid(Server) ->
+    case couch_db:open_int(DbSuffix, [?CTX, sys_db, nologifmissing]) of
```
If the local replicator DB will always be created before this is executed, why use a case clause?

nickva replied:
~Oh you're right. This is unnecessarily defensive. I'll remove it.~
See comment below, I realized why we did it this way

nickva replied:
Just remembered why we did it this way. This is because technically couch_multidb_changes is generic. Replicator app ensures local shard is created. But then this module is trying to behave generically and always check before sending a `resume_scan` message. It did this before in a way  too, we always created a replicator db locally but then still let the scanner find it during the file system scan.

nickva committed:
fixup! Fix whitesapace

nickva committed:
Fix Travis build.

Travis on GH currently copies stuff from this replicator branch into a
replicator source dir from upstream which has an extra test file. That
test file doesn't pass without upstream changes.

Make sure to clear the replicator directory before copying all the
stuff into it.
